src/lib/Simplify.hs

```diff
@@ -174,16 +176,69 @@ simplifyExpr expr = case expr of
       DataCon _ _ con args -> do
         let Abs bs body = alts !! con
         extendR (newEnv bs args) $ simplifyBlock body
-      Variant types label i value -> do
+      Variant (NoExt types) label i value -> do
         let LabeledItems ixtypes = enumerate types
-        let index = fst $ (ixtypes M.! label) !! i
+        let index = fst $ (ixtypes M.! label) NE.!! i
         let Abs bs body = alts !! index
         extendR (newEnv bs [value]) $ simplifyBlock body
       _ -> do
         alts' <- forM alts $ \(Abs bs body) -> do
           bs' <-  mapM (mapM substEmbedR) bs
           buildNAbs bs' $ \xs -> extendR (newEnv bs' xs) $ simplifyBlock body
         emit $ Case e' alts' resultTy'
+  RecordCons left right -> case getType right of
+    RecordTy (NoExt rightTys) -> do
+      -- Unpack, then repack with new arguments (possibly in the middle).
+      left' <- mapM simplifyAtom left
+      rightList <- getUnpacked =<< simplifyAtom right
+      let rightItems = restructure rightList rightTys
+      return $ Record $ left' <> rightItems
+    _ -> emit =<< RecordCons <$> mapM simplifyAtom left <*> simplifyAtom right
+  RecordSplit leftTys@(LabeledItems litems) full -> case getType full of
+    RecordTy (NoExt fullTys) -> do
+      -- Unpack, then repack into two pieces.
+      fullList <- getUnpacked =<< simplifyAtom full
+      let LabeledItems fullItems = restructure fullList fullTys
+          splitLeft fvs ltys = NE.fromList $ NE.take (length ltys) fvs
+          left = M.intersectionWith splitLeft fullItems litems
+          splitRight fvs ltys = NE.nonEmpty $ NE.drop (length ltys) fvs
+          right = M.differenceWith splitRight fullItems litems
+      return $ Record $ Unlabeled $
+        [Record (LabeledItems left), Record (LabeledItems right)]
+    _ -> emit =<< RecordSplit <$> mapM substEmbedR leftTys <*> simplifyAtom full
+  VariantLift leftTys@(LabeledItems litems) right -> case getType right of
+    VariantTy (NoExt rightTys) -> do
+      -- Emit a case statement (ordered by the arg type) that lifts the type.
```

Comment:
Would it make sense to do some optimizations here? For example, if the left cases and right cases have no overlap, then there is nothing complicated that you have to do. Or do you expect those case statements to somehow get simplified away anyway.

Reply:
This is actually a bit subtle, I think. Even if the left and right cases have no overlap in terms of fields, they can still "interleave" with each other in terms of the canonical order of alternatives, because we use alphabetical ordering in the core IR. In the particular special case where the argument is a variant literal, it's true that we can make additional simplifications here, by just changing the first argument to `Variant`. But when the argument isn't a literal (i.e. if it's a `Var`), we can't reach in and change the type of the variant (and, I think, if it remains a `Var` even into the `Imp` lowering, it could actually become a `SumAsProd`, which is much more complicated to lift).

I "cheat" a bit [here](https://github.com/google-research/dex-lang/pull/201/files#diff-31214e9396d485534fa310eb05d1ca3eR240) by first desugaring and then simplifying the resulting case expression, which should optimize away the literal case.

Reply:
Right, the case simplification might be just enough to do the job. But your comment left me wondering a bit -- why would the interleaving of the fields matter here?

Reply:
It doesn't matter for `Variant` atoms themselves, I guess, but it does matter for `SumAsProd` lowering and also for ordering of alternatives for any case statement that matches on that variant, so it's not a simple "cast", if that makes sense.